### ISSUE TITLE
Discontinue use of register_activation_hook() to flush rewrite rules (on multisite)

### ISSUE THREAD

**Issue by westonruter** [Automattic/amp-wp#801]
Currently the rewrite rules get flushed at the `register_activation_hook()`. This works fine for single-site installs. However, it does not work on multisite. For multisite users, they then have to hit the permalinks admin screen for every site in order to flush the rewrite rules. For large networks, this is not feasible. So the plugin should probably do more dynamic checking at runtime if all the rewrite rules are present, and if not, flush them for a given site. This could avoid having to flush rewrite rules after adding support for a custom post type as well. 

For more, see https://jeremyfelt.com/2015/07/17/flushing-rewrite-rules-in-wordpress-multisite-for-fun-and-profit/

Aside: When adding support for pages (#176), the AMP version will probably require using the query param `/?amp` instead of the path `/amp/` endpoint, for the sake of the page rewrite rules. Since AMP cache does not reveal the full URL anyway, having pretty permalinks is somewhat pointless. We couldn't just remove the endpoint, however, without additional work to fix 404s to redirect from `/amp/` to `/?amp`.

**Comment by RavanH** [comment]
Hi Weston, thanks for pointing me here from the wp.org forums :)

In one of my plugins that need rewrite rules to be flushed on first installation and (sometimes) on upgrade, I do something like this to make it work on multisite:

`
define( 'MYPLUGIN_VERSION', 'x.x.x' );
// define or not to define, that is the question 
// meaning: can be done dynamically with get_plugin_data for example
// without throwing yet another constant on the heap... this is just an example

/**
 * Upgrade
 */
function myplugin_upgrade( $old_version ) {
	// FIRST INSTALL
	if ( ! $old_version ) {
		// rewrite rules not available on plugins_loaded
		// and don't flush rules from init as Polylang chokes on that
		// just remove the db option and let WP regenerate them when ready...
		delete_option('rewrite_rules');

		// and make sure rules are regenerated when admin is visited.
		set_transient('myplugin_flush_rewrite_rules','');

		update_option( 'myplugin_version', MYPLUGIN_VERSION );
	}

	// UPGRADE
	// regular upgrade stuff here
}

/**
 * Init
 */
function myplugin_init() {
	// UPGRADE OR FIRST ACTIVATION?
	$old_version = get_option( 'myplugin_version', 0 );
	if ( version_compare( MYPLUGIN_VERSION, $old_version, '>' ) ) {
		myplugin_upgrade( $old_version );
	}

	// regular plugin init stuff here
}
add_action( 'init', 'myplugin_init' );

/**
 * Admin init
 */
function myplugin_admin_init() {
	// CATCH TRANSIENT for flushing rewrite rules after the sitemaps setting has changed
	if (delete_transient('xmlsf_flush_rewrite_rules')) {
		flush_rewrite_rules();
	}

	// regular admin init stuff here
}
add_action( 'admin_init', 'myplugin_admin_init' );
`

The advantage of this is there is no need for calling any get-all-sites-and-looping--through-them-to-do-stuff-for-each-one ... which does not scale for large networks.

Maybe an idea? 

Could even be transposed to core where each "network active" plugin is checked vis-a-vis a per-site "previously network active plugins" list from DB, after which the regular plugin activation hook is called... to simulate single site setup as it where. 

Hmmm, I could make a dedicated proof of concept plugin for this :)

**Comment by westonruter** [comment]
@RavanH Thanks for the comment. I think the approach of checking the DB for whether a flush is needed and then doing the flush when needed is the approach. I don't think using `admin_init` is the right action because it needs to happen whenever the next visitor to a site comes and hits the frontend (and in particular an `/amp/` endpoint). So there needs to be a just-in-time flushing before rewrite rules are matched. In the `amp-options` option array there could be a boolean flag for `rewrite_rules_flushed` which if empty will cause flushing to happen and then the flag is set to prevent re-flushing. A concern there is what to do when the plugin is deactivated, followed by rewrite rules being regenerated, and then the plugin is reactivated: the flag would still be there. So what if the plugin looks at `get_option('rewrite_rules')` to see if there is an AMP endpoint rule endpoint present? That doesn't seem super efficient, but it would be robust.

**Comment by RavanH** [comment]
The thing is that this is basically a WordPress (multisite) issue and not a plugin issue... On plugin level, developers might make the choice to drop register_activation_hook() in favor of some somlution (like I did in my plugin) that is multisite compatible but then again, register_activation_hook() is the official way.

I chose another method in another plugin of mine: on network activation, the default options of the plugin are all set to OFF which means that the plugin functionality is not activated on each site but just becomes available in the admin. Individual site owners can then choose to activate the options which will trigger the needed activation stuff like flushing rules and such.

In this case (AMP plugin) the devs might choose to alter the plugin so that by default, on multisite / network activation NO post types are activated for AMP. This basically would mean no flush is needed and site admins can choose to activate AMP at any point they choose...
